### PR TITLE
ENH: rates, poisson means two-sample comparison  rebased

### PR DESCRIPTION
**Pull request by josef-pkt**
rebased version of #2723 from 2015

original draft version for stackoverflow is in issue #2718

link again to my initial notes
https://github.com/josef-pkt/misc/wiki/Exact-and-Asymptotic-Tests---Rates,-Poisson

### CODE REVIEW

File: statsmodels/stats/rates.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    elif alternative in ['larger', 'l']:
+        mask = stat_space >= stat_sample  - eps
+    elif alternative in ['smaller', 's']:
+        mask = stat_space <= stat_sample + eps
+    else:
+        raise ValueError('invalid alternative')
+
+
+    print('debug')
+    print(stat_sample, mask.sum(), mean1, mean2)
```
Comment (josef-pkt) [inline statsmodels/stats/rates.py:198]: remove print

**Comment by josef-pkt** [comment]
for unit test

a brief search for R: 
all exact tests seem to be conditional tests

base `stats` has poisson.test that uses binomial conditional on the sum of events (uses likelihood to define extreme cases, rejection region, according to Faye not consistent with confint)
https://stat.ethz.ch/R-manual/R-devel/library/stats/html/poisson.test.html

package rateratio.test by Faye has exact test that matches the confidence interval.
https://journal.r-project.org/archive/2010/RJ-2010-008/RJ-2010-008.pdf
https://cran.r-project.org/web/packages/rateratio.test/index.html

package exactci also has poisson with exact pvalue and confint
https://rdrr.io/cran/exactci/man/poisson.exact.html


NCSS/PASS https://ncss-wpengine.netdna-ssl.com/wp-content/themes/ncss/pdf/Procedures/PASS/Tests_for_the_Ratio_of_Two_Poisson_Rates.pdf also quotes Gu

I don't find an E-test in R with google search, 

another article, that I guess didn't look at

Guogen Shan (2015) Exact unconditional testing procedures for comparing two independent Poisson rates, Journal of Statistical Computation and Simulation, 85:5, 947-955, DOI: 10.1080/00949655.2013.855776
sounds like a Monte Carlo comparing exact unconditional methods
(I don't have access right now)

**Comment by josef-pkt** [comment]
this PR fails in lint check, but not in unit tests

File: statsmodels/stats/rates.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+'''
+
+
+from __future__ import division
+import numpy as np
+from scipy import stats
+
+
+# copied from statsmodels.stats.weightstats
+def _zstat_generic2(value, std_diff, alternative):
```
Comment (josef-pkt) [inline statsmodels/stats/rates.py:17]: remove, I only added this to have a standalone script that didn't depend on statsmodels
(for stackoverflow)

File: statsmodels/stats/rates.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    elif alternative in ['larger', 'l']:
+        pvalue = stats.norm.sf(zstat)
+    elif alternative in ['smaller', 's']:
+        pvalue = stats.norm.cdf(zstat)
+    else:
+        raise ValueError('invalid alternative')
+    return zstat, pvalue
+
+
+def poisson_twosample(count1, exposure1, count2, exposure2, ratio_null=1,
```
Comment (josef-pkt) [inline statsmodels/stats/rates.py:34]: rename to `test_poisson_twosample` or `test_poisson_2indep`

File: statsmodels/stats/rates.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+        dist = 'binomial'
+
+    if dist == 'normal':
+        return _zstat_generic2(stat, 1, alternative)
+    else:
+        return stat, pvalue
+
+
+def etest_twopoisson(count1, exposure1, count2, exposure2, ratio_null=1,
```
Comment (josef-pkt) [inline statsmodels/stats/rates.py:136]: rename `etest_poisson_2indep` ?

File: statsmodels/stats/rates.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    elif method in ['wald']:
+        #stat_func = lambda x1, x2:  (x1 - x2 * r_d) / np.sqrt(x1 + x2 * r_d**2)
+        def stat_func(x1, x2):
+            return (x1 - x2 * r_d) / np.sqrt(x1 + x2 * r_d**2)
+        rate2_mle = y2 / n2
+        rate1_mle = y1 / n1
+        rate1 = rate1_mle
+        rate2 = rate2_mle
+
+    # The sampling distribution needs to be based on the null hypotheis
```
Comment (josef-pkt) [inline statsmodels/stats/rates.py:161]: typo hypothesis

File: statsmodels/stats/rates.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        stat = 2 * (np.sqrt(y1 + 3 / 8.) - np.sqrt((y2 + 3 / 8.) * r_d))
+        stat /= np.sqrt(1 + r_d)
+        dist = 'normal'
+    elif method in ['exact-cond', 'cond-midp']:
+        from statsmodels.stats import proportion
+        bp = r_d / (1 + r_d)
+        y_total = y1 + y2
+        stat = None
+        # TODO: why y2 in here and not y1, check definition of H1 "larger"
+        pvalue = proportion.binom_test(y1, y_total, prop=bp, alternative=alternative)
```
Comment (josef-pkt) [inline statsmodels/stats/rates.py:123]: I think this is covered in R stats poisson test

File: statsmodels/stats/tests/test_rates_poisson.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    # testing against two examples in Gu et al
+
+    print('\ntwo-sided')
+    # example 1
+    count1, n1, count2, n2 = 60, 51477.5, 30, 54308.7
+
+    s1, pv1 = poisson_twosample(count1, n1, count2, n2, method='wald')
+    pv1r = 0.000356
+    assert_allclose(pv1, pv1r*2, rtol=0, atol=5e-6)
+    print('wald', s1, pv1 / 2)   # one sided in the "right" direction
```
Comment (josef-pkt) [inline statsmodels/stats/tests/test_rates_poisson.py:20]: remove print statements.
I guess this was just a example script file

File: statsmodels/stats/rates.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+    # The following uses a fixed truncation for evaluating the probabilities
+    # It will currently only work for small counts, so that sf at truncation point
+    # is small
+    # We can make it depend on the amount of truncated sf.
+    # Some numerical optimization or checks for large means need to be added.
+    threshold = 100
+    threshold = stats.poisson.isf(1e-13, max(mean1, mean2))
+    threshold = max(threshold, 100)   # keep at least 100
+    y_grid = np.arange(threshold + 1)
```
Comment (josef-pkt) [inline statsmodels/stats/rates.py:180]: needs options, kwds, for users to change this, see also #6676
I'm not sure it's the same grid, i.e. here it says y_grid

**Comment by josef-pkt** [comment]
I don't find my original script file on my computer. I don't find a notebook either.
Seems the only examples are in the script in #2718

**Comment by josef-pkt** [comment]
comparing with GLM Poisson

wald test doesn't match up exactly, score_test does
`pv2` is using the function in this PR

```
st = res.model.score_test(res.params, exog_extra = [1., 0])
st, st[1][0], pv2
((array([11.67863472]), array([0.00063222]), 1),
 0.0006322190895476891,
 0.0006322188820470972)
```

**Comment by josef-pkt** [comment]
corner case count1 = count2 = 0

In etest, I use a grid that contains the both zero point. This raised RuntimeWarning in the unit test.
I added an small number `eps`  to avoide zero-division, but this case needs to be investigated and get an "official" behavior. (clipping or adding 0.5 like continuity correction in binomial/proportions)

**Comment by josef-pkt** [comment]
bug, missing unit test
In method='exact-cond':  default alternative "2-sided" does not have correct spelling for binom test

values are good, agrees with R stats

```
s2, pv2 = smr.test_poisson_2indep(count1, n1, count2, n2, method='exact-cond',
                                  ratio_null=1., alternative='two-sided')
 -
pv2, 0.0006751826586863219, pv2 - 0.0006751826586863219  # R stats poisson.test
(0.0006751826586863214, 0.0006751826586863219, -5.421010862427522e-19)
```

**Comment by josef-pkt** [comment]
current exact_cond and cond_midp are "minlike"

library `exactci`  Fay

```
> poisson.exact(c(60, 30), c(51477.5, 54308.7), tsmethod="minlike", midp=TRUE)

        Exact two-sided Poisson test (sum of minimum likelihood method), mid-p
        version

data:  c(60, 30) time base: c(51477.5, 54308.7)
count1 = 60, expected count1 = 43.795646313035, p-value =
0.0005572624066191
alternative hypothesis: true rate ratio is not equal to 1
95 percent confidence interval:
 NA NA
sample estimates:
      rate ratio 
2.10999757175465 
```

```
s2, pv2 = smr.test_poisson_2indep(count1, n1, count2, n2, method='cond-midp',
                                  ratio_null=1., alternative='two-sided')
s2, pv2
(None, 0.000557262406619058)
```

**Comment by josef-pkt** [comment]
The two-sided conditional exact test delegates to proportion.binom_test and then to scipy.stats.binom_test

so "minlike" is inherited from scipy.stats, see also #2735

**Comment by josef-pkt** [comment]
another check tomorrow, then merge
I guess more unit tests for one-sided alternatives and value options.

tost equivalence tests can be easily added.
confint left for another day/PR, I haven't checked yet which confint methods have an analytical form and which methods need iterative inversion of hypothesis tests

**Comment by josef-pkt** [comment]
all green
unit test coverage is only missing some "alternative" options for etest, (plus some exceptions or single lines)
